probe/proc/cache.go

```diff
@@ -0,0 +1,139 @@
+package proc
+
+import (
+	"bytes"
+	"errors"
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+)
+
+const (
+	filesCacheLen        = 1024
+	filesCacheExpiration = 60 * time.Second
+)
+
+var (
+	errInvalidCacheEntry = errors.New("invalid cache entry")
+)
+
+type filesCacheEntry struct {
+	file     File
```

Comment:
The justification is to save `open()/close()` cycles and the subsequent syscalls. What I do is to open the file and then rewind (`seek(0)`) every time we want to re-read from those files (as suggested in the [second answer](http://stackoverflow.com/questions/1623731/open-close-strategy-for-proc-pseudo-file/1623931#1623931) in your link, or [here](http://stackoverflow.com/questions/19949283/seek0-on-linux-proc-sys-pseudo-files) and [here](http://stackoverflow.com/questions/4788462/is-it-safe-to-use-lseek-when-reading-from-proc-fs-files-second-time)) (and it is verified by the mere fact that we are really getting updated values and no errors with this mechanism). Files are closed after 60 seconds no matter what (so we can get rid of entries that are not being used).

The only drawback could be the exhaustion of the number of open files, but I think that `filesCacheLen` has a very conservative value (considering that most of the entries will have cached file content and not files handles, so in practice we will keep 1024/3 open files).